src/lparser.c

```diff
@@ -760,136 +760,102 @@ static void open_func (LexState *ls, FuncState *fs, BlockCnt *bl) {
-    return false;
-  }
-#if YK_HL_DEBUG == 2
-  char *leaf = strrchr(filename, '/');
-  if (leaf && strlen(leaf) > 0)
-    leaf++;
-  else
-    leaf = filename;
-#else
-  char *leaf = filename;
-#endif
-  if (line_number != -1) {
-    /* We know the filename and the line number */
-    int off = snprintf(into, into_len, "%s:%d: ", leaf, line_number);
-    if (off == -1) {
-      return false;
-    }
-    char srcl[MAX_READ_SOURCELINE];
-    bool ok = read_source_line(srcl, MAX_READ_SOURCELINE, filename, line_number);
-    if (!ok) {
-      return false;
-    }
-    char *srcl_strip = srcl;
-    /* strip leading whitespace */
-    for (; *srcl_strip == ' ' && *srcl_strip != '\t'; srcl_strip++);
-    /* strip trailing if present */
-    for (char *p = srcl_strip; *p != '\0'; p++) {
-      if ((*p == '\n') || (*p == '\r')) {
-        *p = '\0';
-        break;
-      }
-    }
-    if (snprintf(into + off, into_len - off, "%s", srcl_strip) == -1) {
-      return false;
-    }
-  } else {
-    /* We know the filename, but not the line number */
-    if (snprintf(into, into_len, "%s:?: ?", leaf) == -1) {
-      return false;
-    }
-  }
-  return true;
-}
-
-
+#include "lopnames.h"
 /*
  * Get a Lua-level source line for use in a hot location debug string.
- *
- * Returns `true` if we could get any debug info.
  */
-bool luaG_get_hl_debug_str(Proto *p, int pc, char *into, size_t into_len) {
-  char *filename = NULL;
-  int line_number = -1;
-
-  if (p->source != NULL) {
-    filename = getstr(p->source);
-    if (strlen(filename) > 0 && filename[0] == '@') {
-      filename++; /* strip the leading '@' */
-      line_number = luaG_getfuncline(p, pc);
+char *luaG_ykdebug_str(Proto *f, int pc) {
+  char tmp[256];
+  Instruction i = f->code[pc];
+  const char *opcode = opnames[GET_OPCODE(i)];
+  if (f->source == NULL) {
+    if (snprintf(tmp, sizeof(tmp), "<unknown>:<unknown>:%s", opcode) < 0)
+      goto err;
+  } else {
+    char source[128];
+    if (snprintf(source, sizeof(source), "%.*s", (int) tsslen(f->source), getstr(f->source)) < 0)
+      goto err;
+    if (source[0] == '@') {
+      char *path = source + 1;
+#if YKLUA_DEBUG_STRS == 2
+      char *leaf = strrchr(path, '/');
+      if (leaf != NULL && strlen(leaf) > 0)
+        leaf++;
+      else
+        leaf = path;
+#else
+      char *leaf = path;
+#endif
+      int line_number = luaG_getfuncline(f, pc);
+      if (line_number != -1) {
+        if (snprintf(tmp, sizeof(tmp), "%s:%d: %s", leaf, line_number, opcode) < 0) {
+          goto err;
+        }
+      } else {
+        if (snprintf(tmp, sizeof(tmp), "%s:<unknown>: %s", leaf, opcode) < 0)
+          goto err;
+      }
+    } else {
+      if (snprintf(tmp, sizeof(tmp), "%s:<unknown>: %s", source, opcode) < 0)
+        goto err;
     }
   }
-  return luaG_format_hl_debug_str(filename, line_number, into, into_len);
-}
+  char *dstr;
+  dstr = malloc(strlen(tmp) + 1);
+  lua_assert(dstr != NULL);
+  strcpy(dstr, tmp);
```

Comment:
i'd use `strncpy` here just in case.

Reply:
We've guaranteed above the string is the right size so `strncpy` would obscure that. [Also, it's slightly annoying to access `strncpy` on some platforms!]